Avoid quadratic complexity in log parser

haampie opened the pull request:
TL;DR: there are matching groups trying to match 1 or more occurrences of
something. We don't use the matching group. Therefore it's sufficient to test
for 1 occurrence. This reduce quadratic complexity to linear time.

---

When parsing logs of an mpich build, I'm getting a 4 minute (!!) wait
with 16 threads for regexes to run:

```
In [1]: %time p.parse("mpich.log")
Wall time: 4min 14s
```

That's really unacceptably slow... 

After some digging, it seems a few regexes tend to have `O(n^2)` scaling
where `n` is the string / log line length. I don't think they *necessarily*
should scale like that, but it seems that way. The common pattern is this

```
([^:]+): error
```

which matches `: error` literally, and then one or more non-colons before that. So
for a log line like this:

```
abcdefghijklmnopqrstuvwxyz: error etc etc
```

Any of these are potential group matches when using `search` in Python:

```
abcdefghijklmnopqrstuvwxyz
 bcdefghijklmnopqrstuvwxyz
  cdefghijklmnopqrstuvwxyz
                         ⋮
                        yz
                         z
```

but clearly the capture group should return the longest match.

My hypothesis is that Python has a very bad implementation of `search`
that somehow considers all of these, even though it can be implemented
in linear time by scanning for `: error` first, and then greedily expanding
the longest possible `[^:]+` match to the left. If Python indeed considers
all possible matches, then with `n` matches of length `1 .. n` you
see the `O(n^2)` slowness (i verified this by replacing + with {1,k}
and doubling `k`, it doubles the execution time indeed).

This PR fixes this by removing the `+`, so effectively changing the 
O(n^2) into a O(n) worst case.

The reason we are fine with dropping `+` is that we don't use the
capture group anywhere, so, we just ensure `:: error` is not a match
but `x: error` is.

After going from O(n^2) to O(n), the  𝄹 mpich build log is parsed
in `1.288s`, so about 200x faster.

Just to be sure I've also updated `^CMake Error.*:` to `^CMake Error`,
so that it does not match with all the possible `:`'s in the line.
Another option is to use `.*?` there to make it quit scanning as soon as
possible, but what line that starts with `CMake Error` that does not have
a colon is really a false positive...

sethrj commented:
Rather than dropping those regex, I like your solution of capping lines to 1000 chars (or hell, do 256 so it fits in a cache line...) or my suggestion of using `signal.alarm` (faster and safer than multithreading) to stop processing after a time limit.

haampie commented:
Somehow if I just compile these regexes myself and do standalone `.match` it's not so slow at all :thinking: have to see what's going on here.

haampie commented:
wtf, `search` takes 18s, `match` takes 1e-6 seconds with the same regex.

haampie committed:
Limit the capture length for negative matches in regexes. We don't even
use the match of those anyways. This avoids problems where a string like

very long line: Error: 2 ...
^^^^^^^^^^^^^^

gets matched on "very long line", "ery long line", ... "ine", "ne", "e".

haampie commented:
Edit: comment is mostly redundant now.

<details>

I've taken the liberty to drop the `prefilter`, cause it's just masking the real problem that could stlil be triggered, namely that `search` is 7 orders of magnitude slower than `match` (I don't know why).

And while I was at it I've updated the regexes to reflect whatever cmake ships.

@bernhardkaindl this also means that I may have dropped the custom regexes you've added a while ago for autotools errors. I'm not sure if we actually really want to deviate from the cmake regexes in a class called CTestLogParser :thinking: would you mind contributing your regexes upstream instead? Kitware is usually happy to accept contributions.

With this PR:

```
$ du -sh mpich.log 
15M	mpich.log
$ time spack python -c 'from ctest_log_parser import CTestLogParser; p = CTestLogParser(); p.parse("mpich.log")'
([<ctest_log_parser.BuildError object at 0x7fb308eb6610>, <ctest_log_parser.BuildError object at 0x7fb308eb60a0>, <ctest_log_parser.BuildError object at 0x7fb308eb65e0>, <ctest_log_parser.BuildError object at 0x7fb308eb6040>], [<ctest_log_parser.BuildWarning object at 0x7fb308eb6640>, <ctest_log_parser.BuildWarning object at 0x7fb308eb6670>])

real	0m0.296s
user	0m1.091s
sys	0m0.150s
```

On current develop:

```
$ time spack python -c 'from ctest_log_parser import CTestLogParser; p = CTestLogParser(); p.parse("mpich.log")'
([<ctest_log_parser.BuildError object at 0x7f9df188f070>, <ctest_log_parser.BuildError object at 0x7f9df188f0a0>], [<ctest_log_parser.BuildWarning object at 0x7f9df187d5e0>, <ctest_log_parser.BuildWarning object at 0x7f9df187daf0>, <ctest_log_parser.BuildWarning object at 0x7f9df187da90>, <ctest_log_parser.BuildWarning object at 0x7f9df187db20>, <ctest_log_parser.BuildWarning object at 0x7f9df187db50>, <ctest_log_parser.BuildWarning object at 0x7f9df187db80>, <ctest_log_parser.BuildWarning object at 0x7f9df187dbb0>, <ctest_log_parser.BuildWarning object at 0x7f9df187dbe0>, <ctest_log_parser.BuildWarning object at 0x7f9df187dc10>, <ctest_log_parser.BuildWarning object at 0x7f9df187dc40>, <ctest_log_parser.BuildWarning object at 0x7f9df187dc70>, <ctest_log_parser.BuildWarning object at 0x7f9df187dca0>, <ctest_log_parser.BuildWarning object at 0x7f9df187dcd0>, <ctest_log_parser.BuildWarning object at 0x7f9df187dd00>, <ctest_log_parser.BuildWarning object at 0x7f9df187dd30>, <ctest_log_parser.BuildWarning object at 0x7f9df187dd60>, <ctest_log_parser.BuildWarning object at 0x7f9df187dd90>, <ctest_log_parser.BuildWarning object at 0x7f9df187ddc0>, <ctest_log_parser.BuildWarning object at 0x7f9df187ddf0>, <ctest_log_parser.BuildWarning object at 0x7f9df187de20>, <ctest_log_parser.BuildWarning object at 0x7f9df187de50>, <ctest_log_parser.BuildWarning object at 0x7f9df187de80>, <ctest_log_parser.BuildWarning object at 0x7f9df187deb0>, <ctest_log_parser.BuildWarning object at 0x7f9df187dee0>, <ctest_log_parser.BuildWarning object at 0x7f9df187df10>, <ctest_log_parser.BuildWarning object at 0x7f9df187df40>, <ctest_log_parser.BuildWarning object at 0x7f9df187df70>, <ctest_log_parser.BuildWarning object at 0x7f9df187dfa0>, <ctest_log_parser.BuildWarning object at 0x7f9df187dfd0>, <ctest_log_parser.BuildWarning object at 0x7f9df188f040>, <ctest_log_parser.BuildWarning object at 0x7f9df187dac0>, <ctest_log_parser.BuildWarning object at 0x7f9df188f100>, <ctest_log_parser.BuildWarning object at 0x7f9df188f130>])

real	4m13.460s
```

okay, so the list is somewhat smaller... not sure what that is about. let me check.

</details>

alalazo commented:
> I've taken the liberty to drop the prefilter, cause it's just masking the real problem that could stlil be triggered, namely that search is 7 orders of magnitude slower than match (I don't know why).

IIrc the prefilter was there to avoid the use, which was done before, of regexps triggering recursive backtracking (which are O(2^n) so much worse than 7x slower), see [here](https://swtch.com/~rsc/regexp/regexp1.html). @tgamblin may have some opinions on that.

haampie commented:
Hm, it seems the regexes have quadratic. If I replace `+` with `{1,k}` it's linear in `k` I mean, so O(kn) and with + the worst case is k=n.

Okay, so I was an idiot, `match` is the same as `search` with `^ [regex]`. It only matches from the beginning.

haampie commented:
So, it seems I'm getting the same results ad develop, but with some additional matched warnings. I think I may have added a new regex copied from cmake. For example:

```diff
   >> 5036  ./adio/include/mpipr.h:293:21: warning: 'PMPI_Waitall' accessing 20 bytes in a region of size 0 [-Wstringop-overflow=]
 
+  >> 5038  adio/common/ad_aggregate.c:504:9: note: in expansion of macro 'MPI_Waitall'
+
+  >> 5041  ./adio/include/mpipr.h:293:21: note: referencing argument 3 of type 'MPI_Status *'
+
+  >> 5043  adio/common/ad_aggregate.c:504:9: note: in expansion of macro 'MPI_Waitall'
+
+  >> 5048  /tmp/hstoppel/spack-stage/spack-stage-mpich-3.4.2-ti65xxj23ajcu7oxycpv7ogzdwnabfxm/spack-src/src/include/mpi.h:1597:5: note: in a call to function 'PMPI_Waitall'
+
   >> 5059  ./adio/include/mpipr.h:293:21: warning: 'PMPI_Waitall' accessing 20 bytes in a region of size 0 [-Wstringop-overflow=]
 
+  >> 5061  adio/common/ad_coll_exch_new.c:221:13: note: in expansion of macro 'MPI_Waitall'
+
+  >> 5064  ./adio/include/mpipr.h:293:21: note: referencing argument 3 of type 'MPI_Status *'
+
+  >> 5066  adio/common/ad_coll_exch_new.c:221:13: note: in expansion of macro 'MPI_Waitall'
+
+  >> 5071  /tmp/hstoppel/spack-stage/spack-stage-mpich-3.4.2-ti65xxj23ajcu7oxycpv7ogzdwnabfxm/spack-src/src/include/mpi.h:1597:5: note: in a call to function 'PMPI_Waitall'
+
   >> 5076  ./adio/include/mpipr.h:293:21: warning: 'PMPI_Waitall' accessing 20 bytes in a region of size 0 [-Wstringop-overflow=]
```

haampie commented on lib/spack/external/ctest_log_parser.py:
```diff
@@ -180,45 +145,40 @@ def match(self, text):
+    "[^:]{1,10}: warning",
     "\", line [0-9]+\\.[0-9]+: [0-9]+-[0-9]+ \\([WI]\\)",
+    "^cxx: Warning:",
+    "file: .* has no symbols",
```
Note that this `.*file: .* ...` one was an offender too, starting with `.*`, which is redundant when using `search`.

haampie commented:
Okay, regarding more matched warnings, it seems the `note` related regex was updated by Todd here (probably to make it run faster prior to the prefilter bits, then it got accidentally added when prefilter stuff was added?) https://github.com/spack/spack/pull/7093

The CMake version of the regex is:

```
"([^ :]+):([0-9]+): note:"
```

Todds PR added a `^` which makes it much more strict and not matching anything really:

```
"^([^ :]+):([0-9]+): note:"
```

And then this PR copies the CMake version back with `+` removed to cap the max match length and no match groups because we don't use them:

```
"[^ :]:[0-9]+: note:"
```

haampie commented on lib/spack/external/ctest_log_parser.py:
```diff
@@ -180,45 +145,40 @@ def match(self, text):
     "\\([0-9]*\\): remark #[0-9]*",
     "\".*\", line [0-9]+: remark\\([0-9]*\\):",
     "cc-[0-9]* CC: REMARK File = .*, Line = [0-9]*",
+    "^CMake Warning.*:",
```
```suggestion
    "^CMake Warning",
```

@tgamblin since you added this to the "slow" list, maybe a quick match for `:` might work too through `.*?:` or just dropping looking for `:`, cause what false positives is it guarding against...

haampie commented on lib/spack/external/ctest_log_parser.py:
```diff
@@ -180,45 +145,40 @@ def match(self, text):
+    "[^:]{1,10}: warning [0-9]+:",
     "^\"[^\"]+\", line [0-9]+: [Ww](arning|arnung)",
+    "[^:]{1,10}: warning[ \\t]*[0-9]+[ \\t]*:",
+    "^(Warning|Warnung) ([0-9]+):",
```
I can't believe this one is ever slow... not sure why it was in the slow list. what log has thousands of numbers after a German "Warnung".

sethrj replied:
yeah I think all the "line number" values should probably be 1 through 6 digits top.

haampie commented on lib/spack/external/ctest_log_parser.py:
```diff
@@ -180,45 +145,40 @@ def match(self, text):
+    "[^ :]{1,10} : warning",
+    "[^:]{1,10}: warning",
     "\", line [0-9]+\\.[0-9]+: [0-9]+-[0-9]+ \\([WI]\\)",
+    "^cxx: Warning:",
```
this one is clearly not slow...

haampie commented on lib/spack/external/ctest_log_parser.py:
```diff
@@ -136,6 +98,9 @@ def match(self, text):
     "^ild:([ \\t])*\\(undefined symbol\\)",
+    "[^ :]{1,10} : (error|fatal error|catastrophic error)",
+    "[^:]{1,10}: (Error:|error|undefined reference|multiply defined)",
+    "[^:]{1,10}\\([^\\)]{1,128}\\) ?: (error|fatal error|catastrophic error)",
```
```suggestion
    "[^:]{1,10}\\([^\\)]+\\) ?: (error|fatal error|catastrophic error)",
```
maybe this works too, since `(` and `)` are matched on both sides, so it can't suffer from the many-matches-problem

haampie committed:
Match O(1) at the start

haampie committed:
Avoid potentially many matches until :

haampie commented:
@tldahlgren maybe you're one of the users of the log parsing feature?
